### PR TITLE
Authentication and Cookies

### PR DESCRIPTION
**Pull request by kevinazhu**
1. Added authentication from login form that POSTs the username and password
2. Added cookies that expire after the specified time
3. Checks cookie expiration on both server and through setting browser cookie expiration
4. If refreshing after POST form data, does not keep asking to resend anymore.

NOTE: changed RequestHandler instances to be non-const because of needing to modify the contents of the handler for authentication map

### CODE REVIEW

File: connection.cc [hunk 1]
```diff
@@ -75,25 +73,59 @@ void Connection::handle_request(const boost::system::error_code& error, size_t b
...
-			}
-			else {
-				// have the handler generate a response
-				RequestHandler::Status status = handler->HandleRequest(*request, &response);
-				if (status != RequestHandler::OK)
-				{
-					response.SetStatus(Response::internal_server_error);
-					response.SetBody("500 Internal Server Error");
+				// get the correct handler based on the header
+				RequestHandler* handler = handlers_->Find(request->uri(), &prefix);
```
Comment (dstephan1) [inline connection.cc]: Ahh what happened to all of the consts?
Do you know about mutable? That or you can store the state somewhere outside of the handler.

The constness helped guarantee thread safety. Race conditions on the shared state are now possible.

File: static_handler.cc [hunk 2]
```diff
@@ -24,29 +30,200 @@ const std::unordered_map<std::string,std::string> content_mappings
...
 
-RequestHandler::Status StaticHandler::HandleRequest(const Request& request, Response* response) const
+RequestHandler::Status StaticHandler::HandleRequest(const Request& request, Response* response)
 {
 	BOOST_LOG_TRIVIAL(trace) << "Creating static file response";
 
 	std::string full_path = request.uri();
 	std::string file_path = directory_ + full_path.substr(prefix_.length()); // get file path relative to server
 
+	// Check if cookie exists if authentication
```
Comment (dstephan1) [inline static_handler.cc]: This is some pretty complicated stuff, so complicated that it probably deserves its own class.

File: connection.cc (hunk 1 above)
Comment (kevinazhu) [inline connection.cc]: Yes, true, this is definitely something we could do in the future.

File: static_handler.cc (hunk 2 above)
Comment (kevinazhu) [inline static_handler.cc]: Good idea, we'll keep it as is for this pull request though and will try to refactor later if we have time.

**Pull request merged**